components/tikv_util/src/config.rs

```diff
@@ -1456,68 +1433,88 @@ impl RaftDataStateMachine {
         Ok(())
     }
 
-    /// Enters the `Migrating` state and returns the source directory if a
-    /// migration is needed. Otherwise prepares the target directory for
-    /// opening.
-    pub fn before_open_target(&mut self) -> Option<PathBuf> {
+    /// Returns whether a migration is needed. When it's needed, enters the `Migrating`
+    /// state. Otherwise prepares the target directory for opening.
+    pub fn before_open_target(&mut self) -> bool {
         // Clean up trash directory if there is any.
-        for p in [
-            &self.source,
-            &self.source_staging,
-            &self.target,
-            &self.target_staging,
-        ] {
+        for p in [&self.source, &self.target] {
             let trash = p.with_extension("REMOVE");
             if trash.exists() {
                 fs::remove_dir_all(&trash).unwrap();
             }
         }
-        if Self::data_exists(&self.target_staging) {
-            // InvMigrating -> Completed
-            assert!(!Self::data_exists(&self.source_staging));
-            Self::remove_dir_safe(&self.source);
-            Self::rename_dir_safe(&self.target_staging, &self.target);
-            return None;
-        }
-        if Self::data_exists(&self.source) {
-            // Init -> Migrating
-            assert!(!Self::data_exists(&self.source_staging));
-            Self::rename_dir_safe(&self.source, &self.source_staging);
-        } else if !Self::data_exists(&self.source_staging) {
-            // No source data.
-            return None;
+        if !Self::data_exists(&self.source) {
+            // Recover from Completed state.
+            if self.in_progress_marker.exists() {
+                Self::must_remove(&self.in_progress_marker);
+            }
+            return false;
+        } else if self.in_progress_marker.exists() {
+            // Recover from Migrating state.
+            let real_source = PathBuf::from(fs::read_to_string(&self.in_progress_marker).unwrap());
+            if real_source == self.target {
+                Self::must_remove(&self.source);
+                return false;
+            } else {
+                Self::must_remove(&self.target);
+            }
+        } else {
+            // Init -> Migrating.
+            fs::write(&self.in_progress_marker, self.source.to_str().unwrap()).unwrap();
+            Self::sync_dir(&self.root);
         }
-        Some(self.source_staging.clone())
+        true
     }
 
     /// Exits the `Migrating` state and enters the `Completed` state.
     pub fn after_dump_data(&mut self) {
+        assert!(Self::data_exists(&self.source));
         assert!(Self::data_exists(&self.target));
-        assert!(!Self::data_exists(&self.source));
-        assert!(!Self::data_exists(&self.target_staging));
-        Self::remove_dir_safe(&self.source_staging);
+        Self::must_remove(&self.source); // Enters the `Completed` state.
```

Comment:
I think we may need an extra state here by write a 'complete' state before remove source directory. As remove a non-empty directory is not a atomic operation, if some of the files in source directory are removed but some are not, reinit from this state may cause panic because the source state is corrupted.

Reply:
The remove is atomic by renaming to "xx.TRASH" first.